File: Code for Python version of PRS/AlgoExec_W/GSPWithinCore.py
import numpy as np


class GSPPlus:
    """
    A class to implement the GSPPlus simulation procedure for selecting the best alternative system.
    """

    def __init__(self, alt, params):
        """
        Initialize the GSPPlus object with a list of alternatives and simulation parameters.

        Args:
            alt (list): List of alternative systems to simulate and compare.
            params (list): List of parameters for the GSPPlus procedure, including:
                - n1: Initial number of samples for each alternative.
                - rMax: Maximum number of simulation rounds.
                - beta: Scaling factor for batch size.
                - Additional parameters as needed.
        """
        self.ALT = alt  # List of alternative systems
        self.param = params  # Parameters for the simulation procedure

    def getALT(self):
        """
        Get the list of alternative systems.

        Returns:
            list: The list of alternative systems.
        """
        return self.ALT

    def setALT(self, ALT):
        """
        Set the list of alternative systems.

        Args:
            ALT (list): The new list of alternative systems.
        """
        self.ALT = ALT

    def getParam(self):
        """
        Get the simulation parameters.

        Returns:
            list: The list of parameters for the simulation.
        """
        return self.param

    def setParam(self, param):
        """
        Set the simulation parameters.

        Args:
            param (list): The new list of parameters for the simulation.
        """
        self.param = param

    def stage1Simulate(self):
        """
        Perform stage 1 simulation, collecting initial samples and calculating variance.

        Returns:
            float: The sum of square roots of the variances of the alternatives.
        """
        n1 = int(self.param[0])  # Initial number of samples for each alternative
        sumS = 0  # Variable to store the sum of square roots of variances
        s2 = np.zeros(len(self.ALT))  # Array to store variances of each alternative

        # Collect n1 samples for each alternative and calculate their means
        for i in range(len(self.ALT)):
            samples = np.zeros(n1)  # Array to store samples for each alternative
            mean = 0 # Mean of samples
            self.ALT[i].set_simulation_time(0)
            # Generate n1 samples for the current alternative
            for j in range(n1):
                samples[j] = self.ALT[i].run_system()
                mean += samples[j]/n1
            # Calculate variance for each alternative and update the sum of square roots
            for j in range(n1):
                s2[i] += (samples[j] - mean) ** 2 / (n1 - 1)
            self.ALT[i].set_s2(s2[i])  # Set the variance for each alternative
            sumS += np.sqrt(s2[i])  # Accumulate the sum of square roots of variances

        return sumS

    def set_batch_max(self, sumS):
        """
        Set the batch size and maximum number of simulations for each alternative.

        Args:
            sumS (float): Sum of square roots of the variances of the alternatives.
        """
        n1 = int(self.param[0])  # Initial number of samples
        rMax = int(self.param[1])  # Maximum number of simulation rounds
        beta = int(self.param[2])  # Scaling factor for batch size
        k = int(self.param[5])  # Number of alternatives

        # Calculate and set the batch size and max simulations for each alternative
        for i in range(len(self.ALT)):
            batch = int(np.ceil(beta * (np.sqrt(self.ALT[i].get_s2()) / (sumS / k))))
            self.ALT[i].set_batch(batch)
            self.ALT[i].set_nMax(n1 + rMax * batch)

    def screen(self, eta):
        """
        Screen alternatives based on the eta parameter, eliminating inferior ones.

        Args:
            eta (float): Eta value for screening, used to set the comparison threshold.

        Returns:
            list: The list of alternatives that pass the screening.
        """
        alt = []  # List to store alternatives that pass the screening
        countN = 0  # Count of total samples used
        n1 = int(self.param[0])  # Initial number of samples
        num = len(self.ALT)  # Number of alternatives
        elimination = np.zeros(num, dtype=int)  # Array to track eliminations
        simulate_time = 0  # Total simulation time
        mean = []  # List to store means of alternatives
        c1 = []  # Variance estimate for current samples
        c2 = []  # Variance estimate for maximum samples

        # Perform pairwise comparisons to mark alternatives for elimination
        for i in range(num):
            mean.append(self.ALT[i].get_mean())
            c1.append(self.ALT[i].get_s2() / self.ALT[i].get_num())
            c2.append(self.ALT[i].get_s2() / self.ALT[i].get_nMax())
            for j in range(i):
                y = (mean[i] - mean[j]) / (c1[i] + c1[j])
                a = eta * np.sqrt((n1 - 1) / (c2[i] + c2[j]))
                if y < -a:
                    elimination[i] += 1
                if -y < -a:
                    elimination[j] += 1

        # Collect alternatives that are not eliminated
        for i in range(num - 1, -1, -1):
            if elimination[i] < 1:
                alt.append(self.ALT[i])
            else:
                countN += (self.ALT[i].get_N() + self.ALT[i].get_num())
                simulate_time += self.ALT[i].get_simulation_time()

        # Update the total sample count and simulation time of the first passing alternative
        if alt:
            alt[0].set_N(countN + alt[0].get_N())
            alt[0].set_simulation_time(simulate_time + alt[0].get_simulation_time())
        return alt

    def simulate(self):
        """
        Perform simulations for each alternative based on its batch size.
        """
        for sample_generate in self.ALT:
            for j in range(sample_generate.get_batch()):
                sample_generate.run_system()

    def stage3Simulate(self, h):
        """
        Perform stage 3 simulation to finalize the best alternative.

        Args:
            h (float): Parameter used to calculate the required number of additional samples.

        Returns:
            The best alternative after stage 3 simulation.
        """
        N = 0  # Total number of samples used
        simulate_time = 0  # Total simulation time
        mean = np.zeros(len(self.ALT))  # Array to store means of alternatives

        # Perform additional simulations if needed and update statistics
        for i in range(len(self.ALT)):
            addN = max(0,
                       np.ceil(h * h * self.ALT[i].get_s2() / (self.param[6] * self.param[6])) - self.ALT[i].get_num())
            for j in range(int(addN)):
                self.ALT[i].run_system()
            N += (self.ALT[i].get_N() + self.ALT[i].get_num())
            simulate_time += self.ALT[i].get_simulation_time()
            mean[i] = self.ALT[i].get_mean()

        # Identify the alternative with the highest mean
        maxIndex = np.argmax(mean)
        self.ALT[maxIndex].set_simulation_time(simulate_time)
        self.ALT[maxIndex].set_N(N)
        return self.ALT[maxIndex]

File: Code for Java version of PRS/AlgoExec_W/GSPWithinCore.java
package AlgoExec_W;

import SampleGen.SampleGenerate;

import java.util.ArrayList;
import java.util.List;

/**
 * Class representing the GSP (Generalized Selection Procedure) implementation within a core.
 * This class provides methods for simulating and screening alternatives to select the best one.
 */
public class GSPWithinCore {
    // List of all alternatives
    List<SampleGenerate> ALT;
    // Parameters of the procedure
    double[] param;

    // Getter for the list of alternatives
    public List<SampleGenerate> getALT() {
        return ALT;
    }

    // Setter for the list of alternatives
    public void setALT(List<SampleGenerate> ALT) {
        this.ALT = ALT;
    }

    // Getter for the parameters
    public double[] getParam() {
        return param;
    }

    // Setter for the parameters
    public void setParam(double[] param) {
        this.param = param;
    }

    /**
     * Constructor to initialize the list of alternatives and the parameters.
     *
     * @param ALT   List of SampleGenerate objects representing the alternatives.
     * @param param Array of parameters required by the procedure.
     */
    public GSPWithinCore(List<SampleGenerate> ALT, double[] param) {
        this.ALT = ALT;
        this.param = param;
    }

    /**
     * Perform Stage 1 simulation for all alternatives.
     * This method generates initial samples and calculates the variance for each alternative.
     *
     * @return The sum of the standard deviations of all alternatives.
     */
    public double stage1Simulate() {
        int n1 = (int) param[0]; // Initial number of samples
        double[] s2 = new double[ALT.size()]; // Array to store variances
        double sumS = 0; // Sum of standard deviations

        // Iterate through all alternatives
        for (int i = 0; i < ALT.size(); i++) {
            double[] sample = new double[n1]; // Array to store samples
            double mean = 0; // Mean of samples

            // Generate n1 samples for the current alternative
            for (int j = 0; j < n1; j++) {
                sample[j] = ALT.get(i).runSystem(); // Run the system to get a sample
                mean += sample[j] / n1; // Accumulate the mean
            }

            // Calculate the variance of the samples
            for (int j = 0; j < n1; j++) {
                s2[i] += Math.pow(sample[j] - mean, 2) / (n1 - 1); // Calculate and accumulate the variance
            }
            ALT.get(i).setS2(s2[i]); // Set the variance for the current alternative
            sumS += Math.sqrt(s2[i]); // Accumulate the standard deviation
        }
        return sumS; // Return the sum of standard deviations
    }

    /**
     * Set batch size and maximum number of samples for each alternative based on the average standard deviation.
     *
     * @param sumS The sum of standard deviations of all alternatives.
     */
    public void getBatchrMax(double sumS) {
        int n1 = (int) param[0]; // Initial number of samples
        int rMax = (int) param[1]; // Maximum number of batches
        int beta = (int) param[2]; // Scaling factor for the batch size
        int k = (int) param[5]; // Number of alternatives

        // Iterate through all alternatives
        for (SampleGenerate sampleGenerate : ALT) {
            int batch = (int) Math.ceil(beta * (Math.sqrt(sampleGenerate.getS2()) / (sumS / k))); // Calculate the batch size
            sampleGenerate.setBatch(batch); // Set the batch size
            sampleGenerate.setnMax(n1 + rMax * batch); // Set the maximum number of samples
        }
    }

    /**
     * Screen the alternatives and eliminate those that are unlikely to be the best.
     *
     * @param eta The screening threshold.
     * @return A list of alternatives that pass the screening.
     */
    public List<SampleGenerate> screen(double eta) {
        List<SampleGenerate> alt = new ArrayList<>(); // List to store the remaining alternatives
        int h = ALT.size();
        int countN = 0; // Count of samples for eliminated alternatives
        double simulateTime = 0.0; // Accumulated simulation time for eliminated alternatives
        int n1 = (int) param[0]; // Initial number of samples
        int[] elimination = new int[h]; // Array to track eliminations

        // Compare each pair of alternatives
        double[] Mean = new double[h];
        double[] C1 = new double[h];
        double[] C2 = new double[h];
        Mean[0] = ALT.get(0).getMean();
        C1[0] = ALT.get(0).getS2() / ALT.get(0).getNum();
        C2[0] = ALT.get(0).getS2() / ALT.get(0).getnMax();
        for (int i = 1; i < h; i++) {
            Mean[i] = ALT.get(i).getMean();
            C1[i] = ALT.get(i).getS2() / ALT.get(i).getNum();
            C2[i] = ALT.get(i).getS2() / ALT.get(i).getnMax();
            for (int j = 0; j < i; j++) {
                double y = (Mean[i] - Mean[j]) / (C1[i] + C1[j]); // Calculate the difference in means
                double a = eta * Math.sqrt((n1 - 1) / (C2[i] + C2[j])); // Calculate the threshold
                // Check if the alternatives should be eliminated
                if (y < -a) {
                    elimination[i] += 1;
                }
                if (-y < -a) {
                    elimination[j] += 1;
                }
            }
        }
        // Eliminate alternatives based on the number of eliminations
        for (int i = ALT.size() - 1; i >= 0; i--) {
            if (elimination[i] < 1) {
                alt.add(ALT.get(i)); // Keep the alternative
            } else {
                countN += (ALT.get(i).getN() + ALT.get(i).getNum()); // Accumulate the sample count
                simulateTime += ALT.get(i).getSimulateTime(); // Accumulate the simulation time
            }
        }
        // Update the remaining alternatives with the accumulated values
        alt.get(0).setN(countN + alt.get(0).getN());
        alt.get(0).setSimulateTime(alt.get(0).getSimulateTime() + simulateTime);
        return alt; // Return the list of remaining alternatives
    }

    /**
     * Simulate each alternative based on its batch size.
     */
    public void simulate() {
        // Iterate through all alternatives
        for (SampleGenerate sampleGenerate : ALT) {
            // Generate samples for each alternative based on the batch size
            for (int j = 0; j < sampleGenerate.getBatch(); j++) {
                sampleGenerate.runSystem(); // Run the system to get a sample
            }
        }
    }

    /**
     * Perform Stage 3 simulation to finalize the selection of the best alternative.
     *
     * @param h The Rinott constant used to calculate the number of additional samples.
     * @return The best alternative after the simulation.
     */
    public SampleGenerate stage3Simulate(double h) {
        int N = 0; // Total number of samples
        double[] mean = new double[ALT.size()]; // Array to store means of alternatives
        double simulationTime = 0.0; // Total simulation time

        // Iterate through all alternatives
        for (int i = 0; i < ALT.size(); i++) {
            double addN = Math.max(0, Math.ceil(h * h * ALT.get(i).getS2() / (param[6] * param[6])) - ALT.get(i).getNum()); // Calculate the number of additional samples
            // Generate additional samples for the current alternative
            for (int j = 0; j < addN; j++) {
                ALT.get(i).runSystem(); // Run the system to get a sample
            }
            N += (ALT.get(i).getN() + ALT.get(i).getNum()); // Accumulate the total number of samples
            simulationTime += ALT.get(i).getSimulateTime(); // Accumulate the total simulation time
            mean[i] = ALT.get(i).getMean(); // Store the mean of the current alternative
        }

        // Find the index of the alternative with the maximum mean
        int maxIndex = findMaxIndex(mean);
        ALT.get(maxIndex).setN(N); // Set the total number of samples for the best alternative
        ALT.get(maxIndex).setSimulateTime(simulationTime); // Set the total simulation time for the best alternative
        return ALT.get(maxIndex); // Return the best alternative
    }

    /**
     * Find the index of the maximum value in an array.
     *
     * @param array The array of values.
     * @return The index of the maximum value.
     * @throws IllegalArgumentException if the array is null or empty.
     */
    public static int findMaxIndex(double[] array) {
        if (array == null || array.length == 0) {
            throw new IllegalArgumentException("Array must not be null or empty"); // Validate the input array
        }
        int maxIndex = 0; // Initialize the index of the maximum value
        double maxValue = array[0]; // Initialize the maximum value

        // Iterate through the array to find the maximum value
        for (int i = 1; i < array.length; i++) {
            if (array[i] > maxValue) {
                maxValue = array[i]; // Update the maximum value
                maxIndex = i; // Update the index of the maximum value
            }
        }
        return maxIndex; // Return the index of the maximum value
    }
}

